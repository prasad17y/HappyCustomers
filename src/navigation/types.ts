import {NavigationProp} from '@react-navigation/native';

export type RootStackParamList = {
  UserList: undefined;
  AddUser: undefined;
};

export type RootStackNavigationProp = NavigationProp<RootStackParamList>;
