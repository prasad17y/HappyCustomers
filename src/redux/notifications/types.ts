export type ToastType = 'success' | 'error' | 'info';

export interface ToastPayload {
  message: string;
  type: ToastType;
}

export interface NotificationsState {
  toast: ToastPayload | null;
}

export const SHOW_TOAST = 'SHOW_TOAST';
export const CLEAR_TOAST = 'CLEAR_TOAST';

interface ShowToastAction {
  type: typeof SHOW_TOAST;
  payload: ToastPayload;
}

interface ClearToastAction {
  type: typeof CLEAR_TOAST;
}

export type NotificationsActionTypes = ShowToastAction | ClearToastAction;
