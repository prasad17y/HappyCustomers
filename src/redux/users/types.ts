export interface UsersState {
  isSyncing: boolean;
  syncError: string | null;
  lastSyncTimestamp: string | null;
}

export const SYNC_USERS_REQUEST = 'SYNC_USERS_REQUEST';
export const SYNC_USERS_SUCCESS = 'SYNC_USERS_SUCCESS';
export const SYNC_USERS_FAILURE = 'SYNC_USERS_FAILURE';
export const CLEAR_SYNC_ERROR = 'CLEAR_SYNC_ERROR';

interface SyncUsersRequestAction {
  type: typeof SYNC_USERS_REQUEST;
  payload?: {forceRefresh: boolean};
}

interface SyncUsersSuccessAction {
  type: typeof SYNC_USERS_SUCCESS;
  payload: {didFetch: boolean};
}

interface SyncUsersFailureAction {
  type: typeof SYNC_USERS_FAILURE;
  payload: string;
}

interface ClearSyncErrorAction {
  type: typeof CLEAR_SYNC_ERROR;
}

export type UsersActionTypes =
  | SyncUsersRequestAction
  | SyncUsersSuccessAction
  | SyncUsersFailureAction
  | ClearSyncErrorAction;
