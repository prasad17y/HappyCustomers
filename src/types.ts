export interface UserType {
  id: string;
  name: string;
  email: string | null;
  role: string;
}
