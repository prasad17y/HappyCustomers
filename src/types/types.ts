export interface UserType {
  id: string;
  name: string;
  email: string | null;
  role: string;
}

export enum Role {
  ADMIN = 'ADMIN',
  MANAGER = 'MANAGER',
}

export interface Tab {
  key: string;
  title: string;
}

export interface FilterTab extends Tab {
  role?: Role;
}
