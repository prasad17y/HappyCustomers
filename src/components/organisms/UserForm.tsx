import React from 'react';
import {View, Text, StyleSheet, ScrollView, Keyboard} from 'react-native';
import {FilterTab, Role} from '../../types/types';
import {Formik, FormikHelpers} from 'formik';
import * as yup from 'yup';
import TabsGroup from '../molecules/TabsGroup';
import Button from '../atoms/Button';
import FormTextInput from '../molecules/FormTextInput';

export const validationSchema = yup.object().shape({
  firstName: yup
    .string()
    .trim()
    .matches(/^[a-zA-Z]*$/, 'Name can only contain letters.')
    .required('First Name is required.'),
  lastName: yup
    .string()
    .trim()
    .matches(/^[a-zA-Z]*$/, 'Name can only contain letters.')
    .required('Last Name is required.'),
  email: yup
    .string()
    .transform(value => (value === '' ? undefined : value))
    .email('Please enter a valid email address.')
    .nullable(),
  fullName: yup
    .string()
    .test(
      'max-length',
      'Full name cannot exceed 50 characters.',
      function (_value) {
        const {firstName = '', lastName = ''} = this.parent;
        return (firstName.trim() + ' ' + lastName.trim()).length <= 50;
      },
    ),
});

export interface UserFormValues {
  firstName: string;
  lastName: string;
  email: string;
  role: Role;
  fullName: string;
}

interface UserFormProps {
  initialValues: UserFormValues;
  onSubmit: (
    values: UserFormValues,
    helpers: FormikHelpers<UserFormValues>,
  ) => void | Promise<any>;
  submitButtonText: string;
}

const UserForm: React.FC<UserFormProps> = ({
  initialValues,
  onSubmit,
  submitButtonText,
}) => {
  return (
    <Formik
      initialValues={initialValues}
      validationSchema={validationSchema}
      onSubmit={onSubmit}>
      {({
        handleChange,
        handleBlur,
        handleSubmit,
        values,
        errors,
        touched,
        setFieldValue,
      }) => (
        <ScrollView
          contentContainerStyle={styles.container}
          keyboardShouldPersistTaps="handled">
          <View style={styles.form}>
            <FormTextInput
              value={values.firstName}
              onChangeText={handleChange('firstName')}
              onBlur={handleBlur('firstName')}
              placeholder="First Name"
              error={errors.firstName || errors.fullName}
              touched={touched.firstName}
            />

            <FormTextInput
              value={values.lastName}
              onChangeText={handleChange('lastName')}
              onBlur={handleBlur('lastName')}
              placeholder="Last Name"
              error={errors.lastName || errors.fullName}
              touched={touched.lastName}
            />

            <FormTextInput
              value={values.email}
              onChangeText={handleChange('email')}
              onBlur={handleBlur('email')}
              placeholder="Email"
              keyboardType="email-address"
              autoCapitalize="none"
              error={errors.email}
              touched={touched.email}
            />

            <Text style={styles.label}>User Role</Text>
            <TabsGroup
              tabs={
                [
                  {key: 'admin', title: 'Admin', role: Role.ADMIN},
                  {key: 'manager', title: 'Manager', role: Role.MANAGER},
                ] as FilterTab[]
              }
              selectedIndex={values.role === Role.ADMIN ? 0 : 1}
              onTabPress={(tab: FilterTab) => {
                Keyboard.dismiss();
                setFieldValue('role', tab.role);
              }}
              style={styles.tabsGroup}
            />
          </View>
          <View style={styles.footer}>
            <Button style={styles.submitButton} onPress={() => handleSubmit()}>
              <Text style={styles.submitButtonText}>{submitButtonText}</Text>
            </Button>
          </View>
        </ScrollView>
      )}
    </Formik>
  );
};

const styles = StyleSheet.create({
  container: {
    flexGrow: 1,
  },
  form: {
    flex: 1,
  },
  label: {
    fontSize: 16,
    color: '#666',
    marginBottom: 12,
    marginTop: 16,
  },
  tabsGroup: {
    backgroundColor: '#f0f2f5',
    borderRadius: 22,
    height: 44,
  },
  footer: {
    paddingVertical: 20,
  },
  submitButton: {
    backgroundColor: '#0b5ac2',
    padding: 16,
    borderRadius: 22,
    alignItems: 'center',
  },
  submitButtonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
  },
});

export default React.memo(UserForm);
