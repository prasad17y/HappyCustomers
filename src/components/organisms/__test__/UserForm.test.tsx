import React from 'react';
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react-native';
import UserForm, {UserFormValues, validationSchema} from '../UserForm';
import {Role} from '../../../types/types';

describe('UserForm Organism', () => {
  const initialValues: UserFormValues = {
    firstName: '',
    lastName: '',
    email: '',
    role: Role.MANAGER,
    fullName: '',
  };

  describe('Component Rendering and Interaction', () => {
    it('should render correctly and match the snapshot', () => {
      const tree = render(
        <UserForm
          initialValues={initialValues}
          onSubmit={jest.fn()}
          submitButtonText="Create"
        />,
      );

      fireEvent(tree.getByTestId('userform-tabsgroup-container'), 'layout', {
        nativeEvent: {layout: {width: 200}},
      });

      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should call onSubmit with the correct values for a valid form', async () => {
      const onSubmitMock = jest.fn();
      render(
        <UserForm
          initialValues={initialValues}
          onSubmit={onSubmitMock}
          submitButtonText="Create User"
        />,
      );

      fireEvent(screen.getByTestId('userform-tabsgroup-container'), 'layout', {
        nativeEvent: {layout: {width: 200}},
      });

      // Fill out the form with valid data
      fireEvent.changeText(screen.getByPlaceholderText('First Name'), 'John');
      fireEvent.changeText(screen.getByPlaceholderText('Last Name'), 'Doe');
      fireEvent.changeText(
        screen.getByPlaceholderText('Email'),
        'john.doe@example.com',
      );
      fireEvent.press(screen.getByText('Admin')); // Change role to Admin

      // Press the submit button
      fireEvent.press(screen.getByText('Create User'));

      // Wait for the asynchronous validation and submission to complete
      await waitFor(() => {
        expect(onSubmitMock).toHaveBeenCalledTimes(1);
        // Check that the first argument of the first call has the correct values
        expect(onSubmitMock.mock.calls[0][0]).toEqual({
          firstName: 'John',
          lastName: 'Doe',
          email: 'john.doe@example.com',
          role: Role.ADMIN,
          fullName: '',
        });
      });
    });

    it('should show a validation error for an invalid first name', async () => {
      render(
        <UserForm
          initialValues={initialValues}
          onSubmit={jest.fn()}
          submitButtonText="Create"
        />,
      );

      fireEvent(screen.getByTestId('userform-tabsgroup-container'), 'layout', {
        nativeEvent: {layout: {width: 200}},
      });

      const firstNameInput = screen.getByPlaceholderText('First Name');

      // Type an invalid value
      fireEvent.changeText(firstNameInput, 'John123');

      // Trigger the onBlur event to run validation
      fireEvent(firstNameInput, 'blur', {
        nativeEvent: {text: 'John123'},
        persist: () => {},
        target: {},
      });

      // Wait for the error message to appear
      await waitFor(() => {
        expect(screen.getByText('Name can only contain letters.')).toBeTruthy();
      });
    });
  });

  describe('Validation Schema', () => {
    it('should pass validation for correct data', async () => {
      const validData = {
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@example.com',
      };
      const isValid = await validationSchema.isValid(validData);
      expect(isValid).toBe(true);
    });

    it('should fail validation if first name is missing', async () => {
      const invalidData = {firstName: '', lastName: 'Doe'};
      await expect(validationSchema.validate(invalidData)).rejects.toThrow(
        'First Name is required.',
      );
    });

    it('should fail validation if name contains numbers', async () => {
      const invalidData = {firstName: 'John123', lastName: 'Doe'};
      await expect(validationSchema.validate(invalidData)).rejects.toThrow(
        'Name can only contain letters.',
      );
    });

    it('should fail validation for an invalid email format', async () => {
      const invalidData = {
        firstName: 'Jane',
        lastName: 'Doe',
        email: 'jane.doe@invalid',
      };
      await expect(validationSchema.validate(invalidData)).rejects.toThrow(
        'Please enter a valid email address.',
      );
    });

    it('should fail validation if full name is over 50 characters', async () => {
      const invalidData = {
        firstName: 'ThisIsAVeryLongFirstNameThatWillExceedTheLimit',
        lastName: 'ThisIsAVeryLongLastNameThatWillExceedTheLimit',
      };
      await expect(validationSchema.validate(invalidData)).rejects.toThrow(
        'Full name cannot exceed 50 characters.',
      );
    });
  });
});
