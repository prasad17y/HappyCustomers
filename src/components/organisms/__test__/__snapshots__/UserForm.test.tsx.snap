// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm Organism Component Rendering and Interaction should render correctly and match the snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "flexGrow": 1,
    }
  }
  keyboardShouldPersistTaps="handled"
>
  <View>
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <View
        style={
          {
            "marginBottom": 8,
          }
        }
      >
        <View
          style={
            [
              {
                "borderBottomColor": "#e0e0e0",
                "borderBottomWidth": 1,
              },
              false,
            ]
          }
        >
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="First Name"
            placeholderTextColor="#a0a0a0"
            style={
              [
                {
                  "color": "#191a1a",
                  "fontSize": 16,
                  "paddingVertical": 12,
                },
                undefined,
              ]
            }
            value=""
          />
        </View>
        <View
          style={
            {
              "height": 20,
              "justifyContent": "center",
            }
          }
        />
      </View>
      <View
        style={
          {
            "marginBottom": 8,
          }
        }
      >
        <View
          style={
            [
              {
                "borderBottomColor": "#e0e0e0",
                "borderBottomWidth": 1,
              },
              false,
            ]
          }
        >
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Last Name"
            placeholderTextColor="#a0a0a0"
            style={
              [
                {
                  "color": "#191a1a",
                  "fontSize": 16,
                  "paddingVertical": 12,
                },
                undefined,
              ]
            }
            value=""
          />
        </View>
        <View
          style={
            {
              "height": 20,
              "justifyContent": "center",
            }
          }
        />
      </View>
      <View
        style={
          {
            "marginBottom": 8,
          }
        }
      >
        <View
          style={
            [
              {
                "borderBottomColor": "#e0e0e0",
                "borderBottomWidth": 1,
              },
              false,
            ]
          }
        >
          <TextInput
            autoCapitalize="none"
            keyboardType="email-address"
            onBlur={[Function]}
            onChangeText={[Function]}
            placeholder="Email"
            placeholderTextColor="#a0a0a0"
            style={
              [
                {
                  "color": "#191a1a",
                  "fontSize": 16,
                  "paddingVertical": 12,
                },
                undefined,
              ]
            }
            value=""
          />
        </View>
        <View
          style={
            {
              "height": 20,
              "justifyContent": "center",
            }
          }
        />
      </View>
      <Text
        style={
          {
            "color": "#666",
            "fontSize": 16,
            "marginBottom": 12,
            "marginTop": 16,
          }
        }
      >
        User Role
      </Text>
      <View
        onLayout={[Function]}
        style={
          [
            {
              "backgroundColor": "#f0f2f5",
              "borderRadius": 22,
              "height": 44,
              "padding": 4,
            },
            {
              "backgroundColor": "#f0f2f5",
              "borderRadius": 22,
              "height": 44,
            },
          ]
        }
        testID="userform-tabsgroup-container"
      >
        <View
          style={
            {
              "flex": 1,
              "flexDirection": "row",
              "position": "relative",
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "#e3effa",
                "borderColor": "#0b5ac1",
                "borderRadius": 18,
                "borderWidth": 1,
                "height": "100%",
                "position": "absolute",
                "transform": [
                  {
                    "translateX": 96,
                  },
                ],
                "width": 96,
              }
            }
          />
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "height": "100%",
                "justifyContent": "center",
                "opacity": 1,
                "width": 96,
              }
            }
            testID="userform-tabsgroup-0"
          >
            <Text
              style={
                [
                  {
                    "color": "#666",
                    "fontSize": 16,
                    "fontWeight": "500",
                  },
                  false,
                ]
              }
            >
              Admin
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "height": "100%",
                "justifyContent": "center",
                "opacity": 1,
                "width": 96,
              }
            }
            testID="userform-tabsgroup-1"
          >
            <Text
              style={
                [
                  {
                    "color": "#666",
                    "fontSize": 16,
                    "fontWeight": "500",
                  },
                  {
                    "color": "#0b5ac2",
                    "fontWeight": "500",
                  },
                ]
              }
            >
              Manager
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "paddingVertical": 20,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#0b5ac2",
            "borderRadius": 22,
            "justifyContent": "center",
            "opacity": 1,
            "padding": 16,
          }
        }
      >
        <Text
          style={
            {
              "color": "#fff",
              "fontSize": 18,
              "fontWeight": "bold",
            }
          }
        >
          Create
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
